### PR TITLE
add cpu ,memory calculate for container

### PR DESCRIPTION
**Pull request by HobbyBear**
let sentinel support  container environment.

### CODE REVIEW

**Comment by HobbyBear** [comment]
sentinel 目前计算cpu，内存不支持容器环境，容器环境读取此类信息的文件路径不同，所以我在计算相关指标的代码中加了判断。如果是容器环境，那么我将从sys/fs/cgroup/cpuacct/cpuacct.usage_percpu读取cpu信息，从sys/fs/cgroup/memory/memory.usage_in_bytes获取内存占用信息，然后存到相关的指标变量里，取代了之前只读取宿主机的cpu，内存信息。

**Comment by luckyxiaoqiang** [comment]
Please fix CI.

File: core/system_metric/sys_metric_stat.go [hunk 1]
```diff
@@ -32,6 +38,9 @@ const (
 	NotRetrievedLoadValue     float64 = -1.0
 	NotRetrievedCpuUsageValue float64 = -1.0
 	NotRetrievedMemoryValue   int64   = -1
+	CGroupPath                        = "/proc/1/cgroup"
+	DockerPath                        = "/docker"
```
Comment (sczyh30) [inline core/system_metric/sys_metric_stat.go:42]: What if the container runtime is not Docker? Kubernetes is deprecating Docker as a container runtime after v1.20. Refer https://kubernetes.io/blog/2020/12/02/dont-panic-kubernetes-and-docker/

File: core/system_metric/sys_metric_stat.go [hunk 2]
```diff
@@ -15,7 +15,13 @@
 package system_metric
 
 import (
+	"bufio"
+	"github.com/pkg/errors"
+	"github.com/shirou/gopsutil/v3/cpu"
```
Comment (sczyh30) [inline core/system_metric/sys_metric_stat.go:20]: Please reformat and optimize imports of your code, according to the CI failure.

**Review by sczyh30 — CHANGES_REQUESTED** [review]

**Review by sanxun0325 — COMMENTED** [review]
It might be better to use gopsutil API directly to get the container data.
reference：https://github.com/shirou/gopsutil/blob/master/v3/docker/docker_linux.go

File: core/system_metric/sys_metric_stat.go (hunk 1 above)
Comment (HobbyBear) [reply core/system_metric/sys_metric_stat.go:43]: image
the next line is the path in Kubernetes.

**Comment by HobbyBear** [comment]
> It might be better to use gopsutil API directly to get the container data.
> reference：https://github.com/shirou/gopsutil/blob/master/v3/docker/docker_linux.go

At first, I wanted to use gopsutil too. but i found that gopsutil need to pass container id to get the container metrics. And The get metric path is relative to the host .
image
when we run our code in container, we may hope that the server running like in host. So the path getting the metrics should be the path in container ,not in the host.

**Comment by sanxun0325** [comment]
> > It might be better to use gopsutil API directly to get the container data.
> > reference：https://github.com/shirou/gopsutil/blob/master/v3/docker/docker_linux.go
> 
> At first, I wanted to use gopsutil too. but i found that gopsutil need to pass container id to get the container metrics. And The get metric path is relative to the host .
> image
> when we run our code in container, we may hope that the server running like in host. So the path getting the metrics should be the path in container ,not in the host.

pass container id to get the container metrics,`base `parameter can be passed in. 
https://github.com/shirou/gopsutil/blob/42d667edb6ce16bc36502a3c5b6d0eb7ac8396ec/v3/docker/docker_linux.go#L140

**Comment by HobbyBear** [comment]
> > > It might be better to use gopsutil API directly to get the container data.
> > > reference：https://github.com/shirou/gopsutil/blob/master/v3/docker/docker_linux.go
> > 
> > 
> > At first, I wanted to use gopsutil too. but i found that gopsutil need to pass container id to get the container metrics. And The get metric path is relative to the host .
> > image
> > when we run our code in container, we may hope that the server running like in host. So the path getting the metrics should be the path in container ,not in the host.
> 
> pass container id to get the container metrics,`base `parameter can be passed in. such as `/sys/fs/cgroup `
> https://github.com/shirou/gopsutil/blob/42d667edb6ce16bc36502a3c5b6d0eb7ac8396ec/v3/docker/docker_linux.go#L140

for example ,if i use gopsutil in sentinel and the sentinel is used in a container .The  path to get metrics  is constructed with container_id and base path like this ,/sys/fs/cgroup/cpu/docker/019edf83f18539db17a432fbc39e3a9448435114d7c62920e21373949bc63f34 ,but the path is  relative to the host ,not in a container .if in a container,the path is not found,and our code is running in a container.
the metric file path  of  host is 
image

the metric file path of container is 
image

**Comment by HobbyBear** [comment]
> Please fix CI.

yeap,i fixed it

**Comment by sanxun0325** [comment]
> > It might be better to use gopsutil API directly to get the container data.
> > reference：https://github.com/shirou/gopsutil/blob/master/v3/docker/docker_linux.go
> 
> At first, I wanted to use gopsutil too. but i found that gopsutil need to pass container id to get the container metrics. And The get metric path is relative to the host .
> image
> when we run our code in container, we may hope that the server running like in host. So the path getting the metrics should be the path in container ,not in the host.

The CPU metrics are obtained by passing in the corresponding container ID in the container environment.
`minikube ` is my container.
image

File: core/system_metric/sys_metric_stat.go [hunk 3]
```diff
@@ -156,16 +263,95 @@ func InitCpuCollector(intervalMs uint32) {
...
+	}
+	preContainerCpu, ok := preContainerCpuUsage.Load().(float64)
+
+	if !ok {
+		return 0, errors.New("preContainerCpuUsage load is not float64")
+	}
+	if currentSysCpuTotal-preSysTotalCpu == 0 {
+		return 0, err
+	}
+	return (currentContainerCpuTotal - preContainerCpu) * onlineContainerCpuCount / (currentSysCpuTotal - preSysTotalCpu), err
```
Comment (ZoeShaw101) [inline core/system_metric/sys_metric_stat.go:317]: 没有看到再去更新preContainerCpu preSysTotalCpu了？

Comment (HobbyBear) [reply core/system_metric/sys_metric_stat.go:317]: retrieveAndUpdateCpuStat 方法里我加了更新逻辑，retrieveAndUpdateCpuStat是在定时器里执行

Comment (HobbyBear) [reply core/system_metric/sys_metric_stat.go:317]:
```
func InitCpuCollector(intervalMs uint32) {
	if intervalMs == 0 {
		return
	}
	cpuStatCollectorOnce.Do(func() {
		// Initial memory retrieval.
		retrieveAndUpdateCpuStat()

		ticker := util.NewTicker(time.Duration(intervalMs) * time.Millisecond)
		go util.RunWithRecover(func() {
			for {
				select {
				case <-ticker.C():
					retrieveAndUpdateCpuStat()
				case <-ssStopChan:
					ticker.Stop()
					return
				}
			}
		})
	})
}
```

File: core/system_metric/sys_metric_stat.go [hunk 4]
```diff
@@ -156,16 +263,95 @@ func InitCpuCollector(intervalMs uint32) {
...
+	currentSysCpuTotal, err = getSysCpuUsage()
+	if err != nil {
+		return 0, err
+	}
+	currentContainerCpuTotal, err = getContainerCpuUsage()
+	if err != nil {
+		return 0, err
+	}
+
+	preSysTotalCpu, ok := preSysTotalCpuUsage.Load().(float64)
```
Comment (sanxun0325) [inline core/system_metric/sys_metric_stat.go:305]: when preSysTotalCpuUsage is nil,panic occurs when float64 is converted

File: core/system_metric/sys_metric_stat.go [hunk 5]
```diff
@@ -156,16 +263,95 @@ func InitCpuCollector(intervalMs uint32) {
...
+	if !ok {
+		return 0, errors.New("preSysTotalCpuUsage load is not float64")
+	}
+	preContainerCpu, ok := preContainerCpuUsage.Load().(float64)
```
Comment (sanxun0325) [inline core/system_metric/sys_metric_stat.go:309]: when preContainerCpuUsage is nil,panic occurs when float64 is converted

File: core/system_metric/sys_metric_stat.go [hunk 6]
```diff
@@ -49,6 +58,11 @@ var (
 	TotalMemorySize    = getTotalMemorySize()
 
 	ssStopChan = make(chan struct{})
+
+	isContainer             bool
+	preSysTotalCpuUsage     atomic.Value
+	preContainerCpuUsage    atomic.Value
```
Comment (luckyxiaoqiang) [inline core/system_metric/sys_metric_stat.go:64]: How about `atomic.Float64`?("go.uber.org/atomic")

File: core/system_metric/sys_metric_stat.go [hunk 7]
```diff
@@ -156,16 +263,95 @@ func InitCpuCollector(intervalMs uint32) {
...
+	f, err := os.Open(path)
+	if err != nil {
+		return 0, err
+	}
+	defer f.Close()
+	reader := bufio.NewReader(f)
+	usage, _, err := reader.ReadLine()
+	if err != nil {
+		return 0, err
+	}
```
Comment (luckyxiaoqiang) [inline core/system_metric/sys_metric_stat.go:346]: Repeated codes. Maybe we can extract a method like `readLineFromFile`.

File: core/system_metric/sys_metric_stat.go [hunk 8]
```diff
@@ -156,16 +263,95 @@ func InitCpuCollector(intervalMs uint32) {
...
+	if !ok {
+		return 0, errors.New("preSysTotalCpuUsage load is not float64")
+	}
+	preContainerCpu, ok := preContainerCpuUsage.Load().(float64)
+
+	if !ok {
+		return 0, errors.New("preContainerCpuUsage load is not float64")
+	}
+	if currentSysCpuTotal-preSysTotalCpu == 0 {
+		return 0, err
```
Comment (luckyxiaoqiang) [inline core/system_metric/sys_metric_stat.go:315]: `return 0, nil`?

File: core/system_metric/sys_metric_stat.go (hunk 3 above)
Comment (luckyxiaoqiang) [reply core/system_metric/sys_metric_stat.go:317]: `return (currentContainerCpuTotal - preContainerCpu) * onlineContainerCpuCount / (currentSysCpuTotal - preSysTotalCpu), nil`?

File: core/system_metric/sys_metric_stat.go (hunk 8 above)
Comment (HobbyBear) [reply core/system_metric/sys_metric_stat.go:315]: the method of calculating cpu is similar to docker stats ,refer this 
`https://stackoverflow.com/questions/45196440/cpupercent-metric-from-docker-stats-vs-cgroups`

File: core/system_metric/sys_metric_stat.go [hunk 9]
```diff
@@ -64,6 +78,68 @@ func init() {
...
+	}
+	defer f.Close()
+	reader := bufio.NewReader(f)
+	usage, _, err := reader.ReadLine()
+	if err != nil {
+		return 0, err
+	}
+	perCpuUsages := strings.Fields(string(usage))
+
+	return float64(len(perCpuUsages)), nil
```
[the remark this answers is not in the document]
Comment (HobbyBear) [reply core/system_metric/sys_metric_stat.go:142]: no,it is the cpu count of the container

File: core/system_metric/sys_metric_stat.go [hunk 10]
```diff
@@ -64,6 +78,68 @@ func init() {
...
+	buff := bufio.NewReader(f)
+	for {
+		line, _, err := buff.ReadLine()
+		if err != nil {
+			return false
+		}
+		if strings.Contains(string(line), DockerPath) ||
+			strings.Contains(string(line), KubepodsPath) {
+			return true
+		}
```
[the remark this answers is not in the document]
Comment (HobbyBear) [reply core/system_metric/sys_metric_stat.go:124]: yes,  it is running in docker container or in k8s pods ,if in lxc ,maybe we can add new path /lxc ,but  i have not really run in lxc, i need to practice with that.

File: core/system_metric/sys_metric_stat.go (hunk 6 above)
Comment (HobbyBear) [reply core/system_metric/sys_metric_stat.go]: yeap,fixed it

**Comment by HobbyBear** [comment]
The cpu utilization rate obtained in sentinel is the same as the mechanism displayed by top command, and 4 cores show 400%, so the cpu utilization rate of the process obtained inside the container is still valid, so I closed this merge.

**Pull request closed without merging**